Add support for storing Data-Encryption-Keys in volume metadata

nixpanic opened the pull request:
# Describe what this PR does #

This PR cleans up the KMS implementation and splits it into two parts:

1. KMS for the master encryption key, or Key-Encryption-Key (KEK)
2. DEKStorage as an interface for storing the Data-Encryption-Key (DEK)

By splitting the functionality, it will be easier to implement additional features, including

- encrypting clones and snapshots
- support for KMS's that do not (want) to store DEKs (like Amazon KMS, to limit costs)

## Is there anything that requires special attention ##

The interface is more modular now. An additional KMS-type is added for testing, the SecretsMetadataKMS. This KMS is based on the SecretsKMS that uses a passphrase from the StorageClass secrets for encrypting volumes. The new KMS uses this passphrase as a KEK and stores the DEKs (encrypted) in the metadata of the volume. This emulates how we plan to add support for Amazon KMS.

## Related issues ##

Fixes: #1864

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

nixpanic committed:
util: add DEKStore interface

DEKStore is a new interface that will be used for Storing and Fetching
DEKs. The existing implementations for KMS already function as a
DEKStore, and will be updated to match the interface.

By splitting KMS and DEKStore into two components, the encryption
configuration for volumes becomes more modular. This makes it possible
to implement a DEKStore where the encrypted DEK for a volume is stored
in the metadata of the volume (RBD image).

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: move existing KMS implementations to the DEKStore interface

Use DEKStore API for Fetching and Storing passphrases.

Drop the fallback for the old KMS interface that is now provided as
DEKStore. The original implementation has been re-used for the DEKStore
interface.

This also moves GetCryptoPassphrase/StoreNewCryptoPassphrase functions
to methods of VolumeEncryption.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: add EncryptDEK DecryptDEK to EncryptionKMS interface

By adding these methods, a KMS can explicitly encrypt/decrypt the DEK if
there is no transparent way of doing so.

Hashicorp Vault encrypts the DEK when it it stored, and decrypts it when
fetched. Therefor there is no need to do any encryption in this case.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: implement the DEKStore interface

To accommodate storing DEKs outside a KMS, the DEK can be stored in the
metadata of the volume.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
doc: add implementation details

Update the emcrypted PVC implementation doc with references to the new
EncryptedKMS, DEKStore and VolumeEncryption types.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: introduce VolumeEncryption type

Prepare for grouping encryption related functions together. The main
rbdVolume object should not be cluttered with KMS or DEK procedures.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: configure the DEKStore if the configuration suggests to use metadata

NewVolumeEncryption() will return an indication that an alternative
DEKStore needs to be configured in case the KMS does not support it.

setKMS() will also set the DEKStore if needed, so renaming it to
configureEncryption() makes things clearer.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: use the KMS as DEKStore if it supports it

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: add SecretsMetadataKMS

This new KMS is based on the (default) SecretsKMS, but instead of using
the passphrase for all volumes, the passphrase is used to
encrypt/decrypt a Data-Encryption-Key that is stored in the metadata of
the volume.

CC: Patrick Uiterwijk <puiterwijk@redhat.com> - for encryption guidance
Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
e2e: add test for the SecretsMetadataKMS provider

The new provider should be able to provision and mount an encrypted
volume.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
cleanup: move SecretsKMS in own file

Prepared for an enhanced API to communicate with a KMS and keep the DEK
storage separate. The crypto.go file is already mixed with different
functions, so moving the KMS part into its own file, just like we have
for Hashicorp Vault KMS's.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

humblec commented:
@vasyl-purchel jfyi

humblec commented on internal/rbd/encryption.go:
```diff
@@ -175,21 +184,63 @@ func (rv *rbdVolume) initKMS(ctx context.Context, volOptions, credentials map[st
+
+	rv.encryption, err = util.NewVolumeEncryption(kms)
+
+	// if the KMS can not store the DEK itself, we'll store it in the
```
above is bit unclear to me. In which situation or example the KMS cannot store the DEK itself ? Is it referring to the secret based encryption here ?

nixpanic replied:
Yes, for the `SecretsKMS` base `SecretsMetadataKMS` that is introduced with this PR. Also it is not wished to stored DEKs in Amazon KMS, as the storing of secrets is more costly than storing the DEK in the metadata of the volume.

humblec replied:
oh.. ok.. thanks for confirming.

humblec commented on internal/rbd/encryption.go:
```diff
@@ -175,21 +184,63 @@ func (rv *rbdVolume) initKMS(ctx context.Context, volOptions, credentials map[st
+	return nil
+}
+
+// StoreDEK saves the DEK in the metadata, overwrites any existing contents.
```
nit: may be we can mention rbd image metadata here to make it more clear.

nixpanic replied:
It's the method of the `rbdVolume` type, so it should be clear this is the metadata of the rbd-volume?

humblec replied:
I am fine either way..

humblec commented on internal/rbd/encryption.go:
```diff
@@ -175,21 +184,63 @@ func (rv *rbdVolume) initKMS(ctx context.Context, volOptions, credentials map[st
+}
+
+// FetchDEK reads the DEK from the image metadata.
+func (rv *rbdVolume) FetchDEK(volumeID string) (string, error) {
```
do we foresee a situation where the rbd volume does not exist while we are in StoreDEK/FetchDEK or RemoveDEK?

nixpanic replied:
Not really, it that is needed, we can add the DEK to the rbdVolume struct and store it later. But I hope that added complexity will not be needed.

humblec replied:
Sure.. however isnt it good to validate `rv` is nil before dereferencing it ? or are we doing it in the caller already ?

humblec commented on internal/util/crypto.go:
```diff
@@ -57,56 +52,111 @@ const (
-	return nil
+// EncryptionKMS provides external Key Management System for encryption
+// passphrases storage.
+type EncryptionKMS interface {
```
A wide guess here, do we need a registration method to be clubbed with EncryptionKMS interface ? so that the registration or connection to KMS can be implemented for a KMS with that.

nixpanic replied:
I thought about that too. It would be nice to have, but not something required for this PR. Maybe create an issue for such a cleanup?

humblec replied:
sure. In one way keeping the KMS methods with the connection/registration handle would be good. will open an issue for the same.

vasyl-purchel commented on internal/rbd/rbd_util.go:
```diff
@@ -98,11 +98,10 @@ type rbdVolume struct {
 	Primary             bool
 	ThickProvision      bool
-	KMS                 util.EncryptionKMS
+	encryption          *util.VolumeEncryption
```
so the only concern I have here is that now we're removing "encrypted" flag from the object which means - if any kms configured for the rbdVolume - consider it encrypted... so is it backward compatible with previously created volumes and metadata they have already?

vasyl-purchel replied:
I guess it should be since if kms id is found in volume metadata - we create volume encryption instance I think and go from there... but want to double check really

nixpanic replied:
Yes, it is backwards compatible. The on-disk structure does not change, this is the main `rbdVolume` struct, and it does not have any on-disk representation.

nixpanic committed:
rebase: vendor golang.org/x/crypto and update to latest

The new SecretsMetadataKMS provider encrypts/decrypts DEKs as they are
stored in the metadata of volumes. The encryption/decryption uses
golang.org/x/crypto/scrypt to generate the encryption key from a
passphrase.

While vendoring golang.org/x/crypto, already vendored sub-packages have
been updated.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

humblec commented on docs/design/proposals/encrypted-pvc.md:
```diff
@@ -127,3 +127,20 @@ data:
+
+### Implementation Details
+
+The main components that are used to support encrypted volumes:
```
@nixpanic eventhough we are bit away from supporting cephfs encryption, just wondering how these methods and implementation adjust with when its available, not a concern for now, but good to consider as well.

nixpanic replied:
we don't know yet how a CephFS encryption interface would look like, but I am pretty confident that this split of encryption and DEK-storage does not complicate things for that

humblec replied:
CephFS encryption will be done with `fscrypt` library (https://www.kernel.org/doc/html/v4.19/filesystems/fscrypt.html)  support,  this would be different from the way we make use of rbd image metadata .etc. #1563 ..Considering CephFS encryption is wip, lets move ahead with current version here.

humblec commented on internal/util/secretskms.go:
```diff
@@ -0,0 +1,224 @@
+	Nonce []byte `json:"nonce"`
+}
+
+// EncryptDEK encrypts the plainDEK with a key derived omr the passphrase from
```
looks a typo on `omr` above?

nixpanic replied:
That is corrected now.

humblec commented on internal/util/secretskms.go:
```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2019 The Ceph-CSI Authors.
```
@nixpanic 2019/2021 ?

nixpanic replied:
it is just moving the existing contents to a new file, and adding more things there, not sure it needs an update

humblec replied:
true, I was thinking the copyright can be new as its new file, but current version is also fine.

humblec commented on internal/rbd/encryption.go:
```diff
@@ -175,21 +184,63 @@ func (rv *rbdVolume) initKMS(ctx context.Context, volOptions, credentials map[st
 	return nil
 }
+
+func (rv *rbdVolume) configureEncryption(kmsID string, credentials map[string]string) error {
```
Source code comment make sense ?

nixpanic replied:
Sure, I can add that.

humblec commented on internal/util/crypto.go:
```diff
@@ -57,54 +52,124 @@ const (
-// no need to store is anywhere.
-func (kms SecretsKMS) SavePassphrase(key, value string) error {
-	return nil
+func (ve *VolumeEncryption) RemoveDEK(volumeID string) error {
```
source code comment here too?

nixpanic replied:
yes, will do!

humblec commented on internal/util/secretskms.go:
```diff
@@ -0,0 +1,224 @@
+	passphrase string
+}
+
+func initSecretsKMS(secrets map[string]string) (EncryptionKMS, error) {
```
here too source code comment?

nixpanic replied:
this code moved from one file to an other, it was not modified. but I guess adding comments would be nice too

humblec commented on internal/util/secretskms.go:
```diff
@@ -0,0 +1,224 @@
+	return string(emdData), nil
+}
+
+// DecryptDEK takes the JSON formatted `encryptedMetedataDEK` contents, and the
```
and `it` ( instead of `the`) fetches ... ?

nixpanic replied:
yes, will fix

humblec commented on internal/util/secretskms.go:
```diff
@@ -0,0 +1,227 @@
+// SecretsKMS is default KMS implementation that means no KMS is in use.
+type SecretsKMS struct {
+	integratedDEK
+
```
nit: @nixpanic do we need an extra line here?

humblec commented on internal/util/secretskms.go:
```diff
@@ -0,0 +1,227 @@
+		return nil, err
+	}
+
+	sKMS, ok := eKMS.(SecretsKMS)
```
:+1:

Madhu-1 commented on internal/rbd/encryption.go:
```diff
@@ -54,6 +54,10 @@ const (
+
+	// metadataDEK is the key in the image metadata where the (encrypted)
+	// DEK is stored.
+	metadataDEK = ".rbd.csi.ceph.com/dek"
```
once we mirror the image the image metadata should also get mirroring. I will check this one.

nixpanic replied:
I expect that image metadata is automatically mirrored. But definitely good to check.

Madhu-1 commented on internal/rbd/rbd_journal.go:
```diff
@@ -233,8 +233,8 @@ func (rv *rbdVolume) Exists(ctx context.Context, parentVol *rbdVolume) (bool, er
 	kmsID := ""
-	if rv.Encrypted {
-		kmsID = rv.KMS.GetID()
+	if rv.isEncrypted() {
```
incase if csidriver gets restarted between the rbd image creation and metadata store. are we handling that scenario also? as we are checking the image exists we also need to make sure the metadata is also set.

nixpanic replied:
This behaviour is not changed. The metadata of the image already contains the KMS-ID, it would be problematic if that is missing. We are adding more things in the metadata with the SecretsMetadataKMS, so the potential issue more likely to cause problems when that KMS is in use.

I think #1881 will be used for enhancing logging, and maybe tracking a status/phase of a volume while provisioning. It would be possible to detect incomplete provisioned volumes that way. IMHO it is not something this PR can address.

Madhu-1 commented on internal/util/crypto.go:
```diff
@@ -57,54 +52,129 @@ const (
+// configuration options.
+type integratedDEK struct{}
+
+func (i integratedDEK) requiresDEKStore() DEKStoreType                       { return DEKStoreIntegrated }
+func (i integratedDEK) EncryptDEK(volumeID, plainDEK string) (string, error) { return plainDEK, nil }
+func (i integratedDEK) DecryptDEK(volumeID, encyptedDEK string) (string, error) {
```
make it more readable? move the return to the next line?

nixpanic replied:
sure, will do!

Madhu-1 commented on internal/util/secretskms.go:
```diff
@@ -0,0 +1,227 @@
+
+	emdData, err := json.Marshal(&emd)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert "+
```
question:- returning and error and logging it might leak any key?

nixpanic replied:
no, the keys are never included in the logs

Madhu-1 commented on internal/util/secretskms.go:
```diff
@@ -0,0 +1,227 @@
+	emd := encryptedMetedataDEK{}
+	err = json.Unmarshal([]byte(encryptedDEK), &emd)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert data to "+
```
same here as well. return and logging of error will leak any sensitive information?

nixpanic replied:
no, I do not think so

Madhu-1 commented on internal/util/secretskms.go:
```diff
@@ -0,0 +1,227 @@
+func generateNonce(size int) []byte {
+	nonce := make([]byte, size)
+	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+		panic(err)
```
panic might not be nice once. return error from here?

nixpanic replied:
sure, can do

nixpanic commented:
Rebased and addressed the comments from @Madhu-1, please check again. Thanks!
